File: frontend/lib/api/securityOps.ts
import { api } from './client';

export interface SecurityExecutiveOverview {
  brief_id: string;
  tenant_id: string;
  posture_grade: string;
  overall_health: string;
  composite_risk_score: number;
  open_critical_incidents: number;
  soc_metrics: {
    total_alerts: number;
    critical_high_alerts: number;
    open_incidents: number;
    closed_incidents: number;
    mttd_minutes: number;
    mttc_minutes: number;
    mttr_minutes: number;
    true_positive_rate: number;
    false_positive_rate: number;
    detection_coverage_pct: number;
    containment_success_rate: number;
  };
  domain_breakdown: {
    identity: number;
    ai: number;
    data: number;
    configuration: number;
    integrations: number;
    financial: number;
  };
  executive_recommendations: string[];
  generated_at: string;
}

export interface SecurityEventItem {
  security_event_id: string;
  event_type: string;
  event_category: string;
  timestamp: string;
  tenant_id: string;
  principal_id: string;
  principal_type: string;
  action: string;
  result: string;
  risk_level: string;
  source: string;
  resource_type?: string;
  resource_id?: string;
  ip_metadata: Record<string, any>;
  device_metadata: Record<string, any>;
  evidence: Record<string, any>;
  metadata: Record<string, any>;
}

export interface SecurityDetectionItem {
  detection_id: string;
  tenant_id: string;
  rule_id: string;
  rule_name: string;
  anomaly_type: string;
  severity: string;
  risk_score: number;
  confidence: number;
  mitre_technique_id?: string;
  mitre_tactic?: string;
  description: string;
  evidence_events: string[];
  metadata: Record<string, any>;
  detected_at: string;
}

export interface SecurityAlertItem {
  alert_id: string;
  tenant_id: string;
  title: string;
  description: string;
  severity: string;
  status: string;
  anomaly_type: string;
  mitre_technique_id?: string;
  mitre_tactic?: string;
  risk_score: number;
  confidence_score: number;
  affected_actor_id?: string;
  affected_target_id?: string;
  evidence: Record<string, any>;
  created_at: string;
}

export interface SecurityIncidentItem {
  incident_id: string;
  tenant_id: string;
  title: string;
  description: string;
  severity: string;
  status: string;
  category: string;
  detection_source: string;
  affected_tenants: string[];
  affected_users: string[];
  affected_services: string[];
  affected_resources: string[];
  timeline: Array<Record<string, any>>;
  evidence: Record<string, any>;
  root_cause?: string;
  containment_actions: string[];
  remediation_actions: string[];
  owner?: string;
  resolution?: string;
  created_at: string;
}

export interface InvestigationWorkspaceData {
  incident: SecurityIncidentItem;
  timeline: Array<{
    entry_id: string;
    timestamp: string;
    type: string;
    title: string;
    description: string;
    severity: string;
    source_id: string;
    category: string;
  }>;
  evidence_graph: {
    nodes: Array<{
      id: string;
      label: string;
      type: string;
      severity?: string;
      principal_type?: string;
      resource_type?: string;
    }>;
    edges: Array<{
      source: string;
      target: string;
      relationship: string;
    }>;
    node_count: number;
    edge_count: number;
  };
  hypotheses: Array<{
    id: string;
    title: string;
    confidence: number;
    status: string;
    supporting_evidence: string[];
    contradicting_evidence: string[];
    recommended_verification: string;
  }>;
  recommended_actions: Array<{
    action: string;
    label: string;
    risk: string;
  }>;
}

export interface BlastRadiusData {
  radius_id: string;
  incident_id?: string;
  tenant_id: string;
  overall_impact_score: number;
  affected_users: string[];
  affected_clients: string[];
  affected_tenants: string[];
  affected_services: string[];
  affected_projects: string[];
  affected_documents: string[];
  affected_financial_records: string[];
  affected_integrations: string[];
  affected_ai_agents: string[];
  affected_workflows: string[];
  narrative_summary: string;
  calculated_at: string;
}

export interface ThreatIndicatorItem {
  indicator_id: string;
  tenant_id: string;
  indicator_type: string;
  indicator_value: string;
  threat_category: string;
  severity: string;
  confidence: number;
  source: string;
  reputation: number;
  is_active: boolean;
  observed_at: string;
}

export interface RiskHeatmapItem {
  domain: string;
  category: string;
  risk_score: number;
  severity: string;
  probability: string;
  exposure: string;
  control_strength: string;
}

export interface RunbookItem {
  name: string;
  action_type: string;
  description: string;
}

export const securityOpsApi = {
  getOverview: async () => {
    const res = await api<{ data: SecurityExecutiveOverview }>('/security/overview');
    return res.data;
  },

  getPosture: async () => {
    const res = await api<{ data: any }>('/security/posture');
    return res.data;
  },

  getRisk: async () => {
    const res = await api<{ data: { assessment: any; heatmap: RiskHeatmapItem[] } }>('/security/risk');
    return res.data;
  },

  listEvents: async (limit = 50) => {
    const res = await api<{ data: SecurityEventItem[] }>(`/security/events?limit=${limit}`);
    return res.data;
  },

  ingestEvent: async (payload: {
    event_type: string;
    source: string;
    principal_id: string;
    action: string;
    status: string;
    resource_id?: string;
    ip_address?: string;
    metadata?: Record<string, any>;
  }) => {
    const res = await api<{ data: any }>('/security/events', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
    return res.data;
  },

  listDetections: async () => {
    const res = await api<{ data: SecurityDetectionItem[] }>('/security/detections');
    return res.data;
  },

  listAlerts: async () => {
    const res = await api<{ data: SecurityAlertItem[] }>('/security/alerts');
    return res.data;
  },

  updateAlertStatus: async (alertId: string, status: string) => {
    const res = await api<{ data: any }>(`/security/alerts/${alertId}/status`, {
      method: 'POST',
      body: JSON.stringify({ status }),
    });
    return res.data;
  },

  listIncidents: async () => {
    const res = await api<{ data: SecurityIncidentItem[] }>('/security/incidents');
    return res.data;
  },

  createIncident: async (payload: {
    title: string;
    description: string;
    severity?: string;
    affected_users?: string[];
    affected_services?: string[];
  }) => {
    const res = await api<{ data: any }>('/security/incidents', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
    return res.data;
  },

  getInvestigationWorkspace: async (incidentId: string) => {
    const res = await api<{ data: InvestigationWorkspaceData }>(`/security/investigations/${incidentId}`);
    return res.data;
  },

  getBlastRadius: async (incidentId: string) => {
    const res = await api<{ data: BlastRadiusData }>(`/security/blast-radius/${incidentId}`);
    return res.data;
  },

  listAttackChains: async () => {
    const res = await api<{ data: any[] }>('/security/attack-chains');
    return res.data;
  },

  listRunbooks: async () => {
    const res = await api<{ data: RunbookItem[] }>('/security/runbooks');
    return res.data;
  },

  executeRemediation: async (payload: {
    alert_id: string;
    action_name: string;
    requested_by: string;
    approved_by: string;
    parameters?: Record<string, any>;
  }) => {
    const res = await api<{ data: any }>('/security/remediation/execute', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
    return res.data;
  },

  getEmergencyControls: async () => {
    const res = await api<{ data: Record<string, boolean> }>('/security/emergency-controls');
    return res.data;
  },

  toggleEmergencyControl: async (control: string, enable: boolean, operatorId: string, reason: string) => {
    const res = await api<{ data: any }>('/security/emergency-controls/toggle', {
      method: 'POST',
      body: JSON.stringify({ control, enable, operator_id: operatorId, reason }),
    });
    return res.data;
  },

  listThreatIndicators: async () => {
    const res = await api<{ data: ThreatIndicatorItem[] }>('/security/indicators');
    return res.data;
  },

  queryCopilot: async (query: string, contextId?: string) => {
    const res = await api<{ data: any }>('/security/copilot', {
      method: 'POST',
      body: JSON.stringify({ query, context_id: contextId }),
    });
    return res.data;
  },
};

File: backend/app/schemas/security_ops.py
"""
Pydantic request and response schemas for Phase 46 Security Operations REST API.
"""

from typing import Dict, Any, List, Optional
from datetime import datetime
from pydantic import BaseModel, Field

try:
    from app.security.base import (
        SecuritySeverity,
        AlertStatus,
        IncidentStatus,
        AnomalyType,
        MitreTactic,
        RunbookActionType,
        EmergencySecurityControl,
    )
except ImportError:
    from backend.app.security.base import (
        SecuritySeverity,
        AlertStatus,
        IncidentStatus,
        AnomalyType,
        MitreTactic,
        RunbookActionType,
        EmergencySecurityControl,
    )


# --- Telemetry Ingestion ---

class IngestSecurityEventRequest(BaseModel):
    event_type: str = Field(..., description="Type of event, e.g. auth.login, api.get, agent.tool_execution")
    source: str = Field("auth", description="Source subsystem: auth, api, agent, data, admin, finance, integration")
    principal_id: str = Field("anonymous", description="Actor or user ID")
    action: str = Field("execute", description="Action performed")
    status: str = Field("success", description="success, failure, blocked")
    tenant_id: str = Field("default_tenant", description="Tenant isolation boundary")
    resource_id: Optional[str] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)
    evidence: Dict[str, Any] = Field(default_factory=dict)


# --- Detections & Alerts ---

class SecurityAlertResponse(BaseModel):
    alert_id: str
    tenant_id: str
    title: str
    description: str
    severity: str
    status: str
    anomaly_type: str
    mitre_technique_id: Optional[str] = None
    mitre_tactic: Optional[str] = None
    risk_score: float
    confidence_score: float
    affected_actor_id: Optional[str] = None
    affected_target_id: Optional[str] = None
    evidence: Dict[str, Any]
    created_at: datetime


class UpdateAlertStatusRequest(BaseModel):
    status: AlertStatus


# --- Incidents ---

class CreateIncidentRequest(BaseModel):
    title: str
    description: str
    severity: SecuritySeverity = SecuritySeverity.HIGH
    affected_tenants: List[str] = Field(default_factory=lambda: ["default_tenant"])
    affected_users: List[str] = Field(default_factory=list)
    affected_services: List[str] = Field(default_factory=list)
    affected_resources: List[str] = Field(default_factory=list)
    owner: Optional[str] = None


class UpdateIncidentStatusRequest(BaseModel):
    status: IncidentStatus
    resolution: Optional[str] = None


class SecurityIncidentResponse(BaseModel):
    incident_id: str
    tenant_id: str
    title: str
    description: str
    severity: str
    status: str
    category: str
    detection_source: str
    affected_tenants: List[str]
    affected_users: List[str]
    affected_services: List[str]
    affected_resources: List[str]
    timeline: List[Dict[str, Any]]
    evidence: Dict[str, Any]
    root_cause: Optional[str] = None
    containment_actions: List[str]
    remediation_actions: List[str]
    business_impact: Optional[str] = None
    security_impact: Optional[str] = None
    owner: Optional[str] = None
    resolution: Optional[str] = None
    postmortem: Optional[str] = None
    created_at: datetime
    updated_at: datetime


# --- Remediation & Controls ---

class RemediationExecuteRequest(BaseModel):
    alert_id: str
    action_name: str
    requested_by: str
    approved_by: str
    parameters: Dict[str, Any] = Field(default_factory=dict)
    idempotency_key: Optional[str] = None


class EmergencyControlToggleRequest(BaseModel):
    control: str
    enable: bool
    operator_id: str
    reason: str


# --- Threat Intelligence ---

class ThreatIndicatorCreateRequest(BaseModel):
    indicator_type: str  # IP, DOMAIN, URL, HASH, EMAIL
    indicator_value: str
    threat_category: str
    severity: SecuritySeverity = SecuritySeverity.MEDIUM
    confidence: float = 0.8
    source: str = "MANUAL_INPUT"
    reputation: int = 80


# --- Copilot ---

class SecurityCopilotQueryRequest(BaseModel):
    query: str
    context_id: Optional[str] = None
